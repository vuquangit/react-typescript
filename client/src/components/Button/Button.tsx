import { GET_THEME_CURRENT } from '@/graphql/queries/getThemeCurrent'
import { useQuery } from '@apollo/client'
import React, { FC, forwardRef } from 'react'

import { AppButton } from './Button.styled'
import { IButtonProps } from './Button.types'
const Button: FC<IButtonProps> = forwardRef((props, ref) => {
  const {
    data: { themeMode } = {}
  } = useQuery(GET_THEME_CURRENT)

  // console.log('BUTTON-themeMode', themeMode);

  return (
    <AppButton
      data-testid="app-button"
      ref={ref}
      variant="normal"
      themeMode={themeMode}
      {...props}
    >
      {props.children}
    </AppButton>
  )
})

Button.displayName = 'Button'
export default Button
