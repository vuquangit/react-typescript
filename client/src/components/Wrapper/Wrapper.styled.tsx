/**
 *
 * Styled for Wrapper
 *
 */

import styled from 'styled-components'
import {
  space,
  display,
  border,
  background,
  position,
  flexbox,
  layout,
} from 'styled-system'
import { WrapperProps } from './Wrapper.types'

export const WrapperComponent: React.FC<WrapperProps> = styled('div')<
  WrapperProps
>`
  ${space};
  ${display};
  ${border};
  ${background};
  ${position};
  ${flexbox};
  ${layout};
`
