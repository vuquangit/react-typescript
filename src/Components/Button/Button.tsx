import React, { FC, forwardRef } from 'react'
import { useSelector } from 'react-redux'
import { RootState } from '@/stores/rootReducer'

import { AppButton } from './Button.styled'
import { IButtonProps } from './Button.types'
const Button: FC<IButtonProps> = forwardRef((props, ref) => {
  const themeMode = useSelector((state: RootState) => state.theme.themeMode)

  return (
    <AppButton
      data-testid="app-button"
      ref={ref}
      variant="normal"
      themeMode={themeMode}
      {...props}
    >
      {props.children}
    </AppButton>
  )
})

Button.displayName = 'Button'
export default Button
