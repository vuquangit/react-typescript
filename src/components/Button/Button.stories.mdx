import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'
import Button from '.'
import Space from 'stories/components/styled-system/Space.mdx'
import Styles from 'stories/styles.mdx'

<Meta title="Story/Components/Button" component={Button} />

<Styles />

# Button

With `MDX` we can define a story for `Button` right in the middle of our
Markdown documentation.

export const Template = (args) => <Button {...args} />

<Canvas>
  <Story
    name="Normal"
    args={{ variant: 'normal' }}
    onClick={action('Medium button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Large"
    args={{ variant: 'large' }}
    onClick={action('Large button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Disable"
    args={{ disabled: true }}
    onClick={action('Disable button clicked')}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<div class="subheading">Props</div>

<Space />
