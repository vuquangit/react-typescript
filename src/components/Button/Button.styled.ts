import styled, { css } from 'styled-components'
import { prop, ifProp, switchProp } from 'styled-tools'

import {
  space,
  layout,
  flexbox,
  buttonStyle,
  border,
  typography,
  variant,
  color,
} from 'styled-system'
import { BaseButtonProps } from './Button.types'

export const AppButton = styled('button')<BaseButtonProps>`
  ${color};
  ${space};
  ${layout};
  ${flexbox};
  ${buttonStyle};
  ${border};
  ${typography};

  cursor: ${prop('cursor', 'default')};
  border: 0;

  ${switchProp('themeMode', {
    light: css`
      background-color: #000;
      color: #fff;
    `,
    dark: css`
      background-color: #fff;
      color: #000;
      border: 1px solid #fff;
    `,
  })}

  ${ifProp(
    'disabled',
    css`
      background-color: ${prop('theme.colors.gray', '#999')};
      border-color: ${prop('theme.colors.gray', '#999')};
      pointer-events: none;
    `
  )}

  ${variant({
    variants: {
      normal: {
        // p: '8px',
        fontSize: '16px',
      },
      large: {
        // p: '12px',
        fontSize: '24px',
      },
    },
  })}
`
