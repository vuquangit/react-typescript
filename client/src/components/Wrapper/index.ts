/**
 *
 * Index for Wrapper
 *
 */

import Wrapper from './Wrapper'

export default Wrapper
