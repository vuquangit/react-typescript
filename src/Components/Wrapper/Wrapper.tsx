/**
 *
 * Wrapper
 *
 */

import React, { FC } from 'react'
import { WrapperProps } from './Wrapper.types'
import { WrapperComponent } from './Wrapper.styled'
import { ThemeProvider } from 'styled-components'
import * as theme from 'theme/theme'

const Wrapper: FC<WrapperProps> = (props: WrapperProps) => {
  return (
    <ThemeProvider theme={theme}>
      <WrapperComponent {...props} />
    </ThemeProvider>
  )
}

export default Wrapper
