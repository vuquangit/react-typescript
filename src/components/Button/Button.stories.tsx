import React from 'react'
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0'
import Wrapper from '@/test/supports/AppWrapper'

// import {
//   Title,
//   Subtitle,
//   Description,
//   Primary,
//   ArgsTable,
//   Stories,
//   PRIMARY_STORY,
// } from '@storybook/addon-docs/blocks'

import { Button } from '.'
import { IButtonProps } from './Button.types'

export default {
  title: 'Story/Components/Button',
  component: Button,
  argTypes: {
    label: {
      control: {
        type: 'text',
      },
    },
    variant: {
      control: {
        type: 'select',
        options: ['normal', 'large'],
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
} as Meta

const Template: Story<IButtonProps> = (args) => (
  <Wrapper themeMode="light">
    <Button {...args}>Button story</Button>
  </Wrapper>
)

export const Normal = Template.bind({})
Normal.args = {
  label: 'Normal button',
  variant: 'normal',
}

export const Large = Template.bind({})
Large.args = {
  label: 'Large button',
  variant: 'large',
}

export const Disable = Template.bind({})
Disable.args = {
  disabled: true,
}
