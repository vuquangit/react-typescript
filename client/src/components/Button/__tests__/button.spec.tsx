import React from 'react'
import { render } from '@testing-library/react'
import Wrapper from '@/test/supports/Wrapper'

import Button from '..'

describe('App Button', () => {
  it('Test Button with light mode', async () => {
    const { container, getByTestId } = render(
      <Wrapper themeMode="light">
        <Button size="large">Primary</Button>
      </Wrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'background-color': '#000',
      color: '#fff',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })

  it('Test Button with dark mode', async () => {
    const { container, getByTestId } = render(
      <Wrapper themeMode="dark">
        <Button size="large">Primary</Button>
      </Wrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'background-color': '#fff',
      color: '#000',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })

  it('Test Button disabled', async () => {
    const { container, getByTestId } = render(
      <Wrapper themeMode="light">
        <Button disabled>Primary</Button>
      </Wrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'pointer-events': 'none',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })
})
