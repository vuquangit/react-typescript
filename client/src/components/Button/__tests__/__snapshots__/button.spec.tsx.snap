// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Button Test Button disabled 1`] = `
.c0 {
  cursor: default;
  background-color: #000;
  color: #fff;
  background-color: #999;
  border-color: #999;
  pointer-events: none;
  font-size: 16px;
}

<div>
  <button
    class="c0"
    data-testid="app-button"
    disabled=""
  >
    Primary
  </button>
</div>
`;

exports[`App Button Test Button with dark mode 1`] = `
.c0 {
  width: large;
  height: large;
  cursor: default;
  background-color: #fff;
  color: #000;
  font-size: 16px;
}

<div>
  <button
    class="c0"
    data-testid="app-button"
  >
    Primary
  </button>
</div>
`;

exports[`App Button Test Button with light mode 1`] = `
.c0 {
  width: large;
  height: large;
  cursor: default;
  background-color: #000;
  color: #fff;
  font-size: 16px;
}

<div>
  <button
    class="c0"
    data-testid="app-button"
  >
    Primary
  </button>
</div>
`;
